idx/memory/tag_query.go

```diff
@@ -0,0 +1,408 @@
+					key:   e.key,
+					value: "^.+",
+				})
+			} else {
+				q.match = append(q.match, kv{
+					key:   e.key,
+					value: "^.+",
+				})
+			}
+		} else {
+			// always anchor all regular expressions at the beginning
+			if (e.operator == MATCH || e.operator == NOT_MATCH) && len(e.value) > 0 && e.value[0] != byte('^') {
+				e.value = "^" + e.value
+			}
+
+			switch e.operator {
+			case EQUAL:
+				q.equal = append(q.equal, e.kv)
+			case NOT_EQUAL:
+				q.notEqual = append(q.notEqual, e.kv)
+			case MATCH:
+				q.match = append(q.match, e.kv)
+			case NOT_MATCH:
+				q.notMatch = append(q.notMatch, e.kv)
+			}
+		}
+	}
+
+	return q, nil
+}
+
+func (q *TagQuery) getInitialByMatch(index TagIndex) (int, TagIDs, error) {
+	resultSet := make(TagIDs)
+	lowestCount := math.MaxUint32
+	startId := 0
+
+	// choose key that has the smallest number of values because we want to
+	// reduce the number of matches and start with the smallest possible resultSet.
+	// the smaller the resultSet the less time will have to be spent to further
+	// filter its values in the following expressions.
+	for i := range q.match {
+		l := len(index[q.match[i].key])
+		if l < lowestCount {
+			lowestCount = l
+			startId = i
+		}
+	}
+
+	// shortcut if pattern is ^.+ (f.e. expression "key!=~" will be translated to "key=~^.+")
+	if q.match[startId].value == "^.+" {
+		for _, ids := range index[q.match[startId].key] {
+			for id := range ids {
+				resultSet[id] = struct{}{}
+			}
+		}
+		return startId, resultSet, nil
+	}
+
+	re, err := regexp.Compile(q.match[startId].value)
+	if err != nil {
+		return 0, nil, errInvalidQuery
+	}
+
+	for v, ids := range index[q.match[startId].key] {
+		if !re.MatchString(v) {
+			continue
+		}
+
+		for id := range ids {
+			resultSet[id] = struct{}{}
+		}
+	}
+	return startId, resultSet, nil
+}
+
+func (q *TagQuery) getInitialByEqual(index TagIndex) (int, TagIDs) {
+	resultSet := make(TagIDs)
+	lowestCount := math.MaxUint32
+	startId := 0
+
+	// choose key-value combo that has the smallest number of series IDs because
+	// we want to start with the smallest possible resultSet.
+	// the smaller the resultSet the less time will have to be spent to further
+	// filter its values in the following expressions.
+	for i := range q.equal {
+		l := len(index[q.equal[i].key][q.equal[i].value])
+		if l < lowestCount {
+			lowestCount = l
+			startId = i
+		}
+	}
+
+	// copy the map, because we'll later delete items from it
+	for k, v := range index[q.equal[startId].key][q.equal[startId].value] {
+		resultSet[k] = v
+	}
+
+	return startId, resultSet
+}
+
+func (q *TagQuery) filterByEqual(skipEqual int, resultSet TagIDs, index TagIndex) {
+	for i, e := range q.equal {
+		if i == skipEqual {
+			continue
+		}
+
+		for id := range resultSet {
+			if _, ok := index[e.key][e.value][id]; !ok {
+				delete(resultSet, id)
+			}
+		}
+	}
+}
+
+func (q *TagQuery) filterByNotEqual(resultSet TagIDs, index TagIndex, byId map[string]*idx.Archive) {
```

Comment:
I might be missing a case here, but why doesn't this look like `filterByEqual` but with `ok` instead of `!ok`? in the innermost loop?

Reply:
you're absolutely right: https://github.com/grafana/metrictank/pull/729/commits/93942e3f4ef895932699800549c7ec7eb0491724